HTML: (empty) href attribute on knowl links

Alex-Jordan opened the pull request:
Local accessibility staff were having screen reader issues with knowl links that did not identify as something the user can click. That led to this.

This PR adds `href=""` to knowl anchor elements. Without an `@href`, an anchor element does not default to having role "link", which is what the screen reader wants to identify the anchor as clickable (either that, or role "button"). Note we could put `role="link"` on this instead, but should not do both.

Things about this may be deceptive. If I visit a (current) page with Firefox and use the inspector tool, I see `href=""`. But it's something the browser is adding while building the DOM, it's not currenlty in the PTX HTML. And apparently, in some browser/screen-reader combinations, that is not good enough for the screen reader to understand the role is "link".

davidfarmer commented:
I thought we had already added the empty href, for the same reason.

I am torn between the simplicity of adding the empty href,
and the fact we are using a side-effect to accomplish the intent.

It won't break knowls, if that is the question.  Or if it does
(I have not tested) I should be able to fix it.

On Mon, 11 Oct 2021, Rob Beezer wrote:

> 
> @davidfarmer will need to comment on this one. Any red flags to how knowls work? Or more precisely, how
> they are recognized?
> 
> Why not @role='link'? That strikes me as a bit less mysterious, and less likely to cause confusion with
> "real" anchor elements.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.[AABTULHHCTHRBCSCMAYREKTUGNIK7A5CNFSM5FY75LP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGO
> HAHBT2I.gif]
> 
> 
>

Alex-Jordan commented:
I didn't see any badness with knowl behavior after building the sample article with this.

> I thought we had already added the empty href, for the same reason.

I thought so as well. I do wonder if I was misled by the browser inspector inserting an `href=""` into the DOM even though it is not in the source.

That's the same reason why I lean toward `href=""` rather than `role="link"`. Because if we end up having both (we added `role="link"` and the browser put `href=""` in the DOM) then that runs contrary to an ARIA role principle: never redundantly put a role equal to the default role.  Officially, `a[href]` has default role "link". While `a[not(href)]` does not*. It's all a bit more confusing when the source says one thing and the DOM says something else, and it's unclear at which stage the screen reader and the accessibility tree come into play.

*Also, what is the purpose of `a[not(href)]`? Googling tells me that this is for indicating that something *would* be a link under different conditions. Like this is a link to see some content, but only after you log in to the site. So it is still styled to look like a link, but it's not a link. It sounds like another reason why we really should have the `href=""` in there.

Alex-Jordan commented:
Force pushed. 

Unsure if all the repeat commentary is desired.

rbeezer commented:
More commentary is better than less.  ;-)

Rebuilding website now, give it 15 minutes.  Testing on Section 4 of the sample article does not reveal any harm to output.  But please double-check!
